plugins/catalog-backend-module-puppetdb/src/providers/PuppetDbEntityProvider.test.ts

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { TaskInvocationDefinition, TaskRunner } from '@backstage/backend-tasks';
+import { ConfigReader } from '@backstage/config';
+import { getVoidLogger } from '@backstage/backend-common';
+import { PuppetDbEntityProvider } from './PuppetDbEntityProvider';
+import { EntityProviderConnection } from '@backstage/plugin-catalog-backend';
+import * as p from '../puppet/read';
+import { ANNOTATION_PUPPET_CERTNAME } from '../puppet';
+import {
+  ANNOTATION_LOCATION,
+  ANNOTATION_ORIGIN_LOCATION,
+} from '@backstage/catalog-model/';
+import { DEFAULT_ENTITY_OWNER, ENDPOINT_NODES } from '../puppet/constants';
+
+const logger = getVoidLogger();
+
+jest.mock('../puppet/read', () => ({
+  __esModule: true,
+  default: jest.fn(),
+  readPuppetNodes: null,
+}));
+
+class PersistingTaskRunner implements TaskRunner {
+  private tasks: TaskInvocationDefinition[] = [];
+
+  getTasks() {
+    return this.tasks;
+  }
+
+  run(task: TaskInvocationDefinition): Promise<void> {
+    this.tasks.push(task);
+    return Promise.resolve(undefined);
+  }
+}
+
+describe('PuppetEntityProvider', () => {
+  const config = new ConfigReader({
+    catalog: {
+      providers: {
+        puppetdb: {
+          host: 'http://puppetdb:8080',
+          schedule: {
+            frequency: {
+              minutes: 10,
+            },
+            timeout: {
+              minutes: 10,
+            },
+          },
+        },
+      },
+    },
+  });
+
+  describe('where there are no nodes', () => {
+    beforeEach(() => {
+      // @ts-ignore
```

Comment:
Agreed regarding `// @ts-ignore`. However, looking into more examples in this repository, I see we are using `jest.spyOn` in a lot of places, which does the job and does not require `jest.spyOn`.

To avoid major refactoring (converting this into a class), I've switched to `jest.spyOn`. Actually, this even allowed catching a bug in the tests (the mocked entity structure was actually not correct) 🙂. Hope it's OK with you.